tensorflow/compiler/mlir/hlo/lib/Transforms/copy_removal.cc

```diff
@@ -0,0 +1,186 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "mlir-hlo/Analysis/userange_analysis.h"
+#include "mlir-hlo/Dialect/mhlo/IR/lhlo_ops.h"
+#include "mlir-hlo/Transforms/PassDetail.h"
+#include "mlir-hlo/Transforms/passes.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/BufferUtils.h"
+
+namespace mlir {
+
+namespace {
+
+class CopyRemoval : BufferPlacementTransformationBase {
+public:
+  explicit CopyRemoval(Operation *op)
+      : BufferPlacementTransformationBase(op), userange(op, allocs, aliases),
+        dominators(op) {}
+
+  void removeCopy(Operation *op) {
+    // A set with the copy Operation to process.
+    llvm::SetVector<Operation *> toProcess;
+    fillProcessSet(toProcess);
+
+    while (!toProcess.empty()) {
+      Operation *currentOp = toProcess.pop_back_val();
+
+      // Cast the Operation and get the Source and Target.
+      auto copyOpInterface = dyn_cast<CopyOpInterface>(currentOp);
+      Value copySource = copyOpInterface.getSource();
+      Value copyTarget = copyOpInterface.getTarget();
+
+      // Get the UserangeIntervals.
+      UseInterval::Vector sourceInterval =
+          *userange.getUserangeInterval(copySource).getValue();
+      UseInterval::Vector targetInterval =
+          *userange.getUserangeInterval(copyTarget).getValue();
+
+      // Compute the intersection.
+      UseInterval::intervalIntersect(sourceInterval, targetInterval);
+
+      // If the sourceInterval contains more than one UseInterval, there are
+      // multiple operations that intersect. The sourceInterval must have at
+      // least one UseInterval that contains the copyOp.
+      if (sourceInterval.size() > 1)
+        continue;
+
+      // Check if all Operations inside the intersection are part of the copyOp.
+      if (!checkAncestor(currentOp, *sourceInterval.begin()))
+        continue;
+
+      // Check if the currentOp dominates all uses of the copyTarget.
+      if (!checkDominance(currentOp, copyTarget))
+        continue;
+
+      // Replace all uses of the target with the source.
+      copyTarget.replaceAllUsesWith(copySource);
+
+      // Merge the Useranges.
+      UseInterval::intervalMerge(sourceInterval, targetInterval);
```

Comment:
Is this meant to update? `sourceInterval` is a copy.

Reply:
Something went wrong here. Should be fixed with the next commit.